### PR TITLE
Pre-v0.4.3

### PR DESCRIPTION
**Pull request by LightArrowsEXE**

### CODE REVIEW

**Comment by LightArrowsEXE** [comment]
@Ichunjo @Setsugennoao are there any other functions in (public) packages that I can replace in the `util` submodule? I think we're at the point where I'd rather depend on other packages that keep this updated than have to maintain my own utils.

File: lvsfunc/deblock.py [hunk 1]
```diff
@@ -40,7 +40,8 @@ def autodb_dpir(clip: vs.VideoNode, edgevalue: int = 24,
     If that happens, you can perform a denoise on the original clip (maybe even using dpir's denoising mode)
     and grab the brightest pixels from your two clips. That should return a perfectly fine clip.
 
-    Thanks Vardë, louis, setsugen_no_ao!
+    Thanks `Vardë <https://github.com/Ichunjo>`_, `louis <https://github.com/tomato39>_`,
+    `setsugen_no_ao <https://github.com/Setsugennoao>`_!
```
Comment (Setsugennoao) [inline lvsfunc/deblock.py:44]: Setsugen no ao please <3

File: lvsfunc/deblock.py [hunk 2]
```diff
@@ -151,7 +155,7 @@ def dpir(
     matrix: Matrix | int | None = None, cuda: bool | Literal['trt'] | None = None, i444: bool = False,
     tiles: int | Tuple[int, int] | None = None, overlap: int | Tuple[int, int] | None = 8,
     zones: List[Tuple[Range | List[Range] | None, SupportsFloat | vs.VideoNode | None]] | None = None,
-    fp16: bool | None = None, num_streams: int = 1, device_id: int = 0, kernel: Kernel | str = Catrom()
+    fp16: bool | None = None, num_streams: int = 1, device_id: int = 0, kernel: Kernel | str = Bicubic(b=0, c=0.5)
```
Comment (Setsugennoao) [inline lvsfunc/deblock.py:158]: Why? Those are even the default values for Bicubic itself

File: lvsfunc/deblock.py [hunk 3]
```diff
@@ -165,44 +169,63 @@ def dpir(
     Converts to RGB -> runs DPIR -> converts back to original format, and with no subsampling if ``i444=True``.
     For more information, see `the original DPIR repository <https://github.com/cszn/DPIR>`_.
 
+    Thanks `setsugen_no_ao <https://github.com/Setsugennoao>`_!
```
Comment (Setsugennoao) [inline lvsfunc/deblock.py:172]: Setsugen no ao

File: lvsfunc/deblock.py [hunk 4]
```diff
@@ -165,44 +169,63 @@ def dpir(
...
+    :param fp16:                    Represent the clip with 16f tensors instead of 32f for a speedup, but it's useless—
+                                    and may even harm performance—when enabled with a device that doesn't support it.
+    :param num_streams:             Number of concurrent CUDA streams to use. Increase if GPU isn't getting saturated.
+    :param device_id:               Specifies the GPU device id to use.
+    :param kernel:                  py:class:`vskernels.Kernel` object used for conversions between YUV <-> RGB.
+                                    This can also be the string name of the kernel
+                                    (Default: py:class:`vskernels.Bicubic(0, 0.5)`).
+
+    :return:                        Deblocked or denoised clip in either the given clip's format or YUV444PS.
+
```
Comment (Setsugennoao) [inline lvsfunc/deblock.py:223]: https://discord.com/channels/@me/856164891098742804/986762365141942442 👀

File: lvsfunc/dehalo.py [hunk 5]
```diff
@@ -32,6 +34,9 @@ def bidehalo(clip: vs.VideoNode, ref: vs.VideoNode | None = None,
     """
     Dehaloing function that uses bilateral and BM3D to remove bright haloing around edges.
 
+    .. warning::
+        | This function has been deprecated! It will removed in a future commit.
```
Comment (Setsugennoao) [inline lvsfunc/dehalo.py:38]: will be

File: lvsfunc/dehalo.py [hunk 6]
```diff
@@ -203,31 +212,33 @@ def fine_dehalo(clip: vs.VideoNode, ref: vs.VideoNode | None = None,
...
+                                    Must be greater than 1.
+    :param show_mask:               Return mask clip. Valid options are 1–7.
+    :param planes:                  Specifies which planes will be processed.
+                                    Any unprocessed planes will be simply copied.
+
+    :return:                        Dehalo'd clip or halo mask clip.
+
+    :raises ModuleNotFoundError:    Dependencies are missing.
     """
     warnings.warn("fine_dehalo: 'This function is deprecated in favor of `vsdehalo.fine_dehalo`! "
```
Comment (Setsugennoao) [inline lvsfunc/dehalo.py:243]: You put a warning here but in bidehalo you only put a warning in the docstrings, what's the way to deprecate a function then?

File: lvsfunc/deinterlace.py [hunk 7]
```diff
@@ -232,6 +240,11 @@ def _set_paths() -> None:
             p.parent.mkdir(parents=True)
 
     if not (tfm_f.exists() and tdec_f.exists()):
+        warnings.warn("tivtc_vfr: 'When calculating the matches and metrics for the first time, "
```
Comment (Setsugennoao) [inline lvsfunc/deinterlace.py:243]: Should open an issue for this tbh, if you do, assign it to me

File: lvsfunc/scale.py [hunk 8]
```diff
@@ -288,7 +292,7 @@ def descale(clip: vs.VideoNode,
 
 def ssim_downsample(clip: vs.VideoNode, width: int | None = None, height: int = 720,
                     smooth: float | VSFunction = ((3 ** 2 - 1) / 12) ** 0.5,
-                    kernel: Kernel | str = Catrom(), gamma: bool = False,
+                    kernel: Kernel | str = Bicubic(b=0, c=1/2), gamma: bool = False,
```
Comment (Setsugennoao) [inline lvsfunc/scale.py:295]: Why did you change all Catrom occurrencies to Bicubic(b=0, c=0.5) or Bicubic(b=0, c=1/2)? Really xd

File: lvsfunc/util.py [hunk 9]
```diff
@@ -59,6 +56,9 @@ def quick_resample(clip: vs.VideoNode,
 
     :return:            Filtered clip in original depth.
     """
+    warnings.warn("quick_resample: 'This function will be either reworked or removed in a future version!'",
```
Comment (Setsugennoao) [inline lvsfunc/util.py:59]: 💣 💣 💣 
   Nuke it

File: lvsfunc/util.py [hunk 10]
```diff
@@ -231,31 +185,28 @@ def replace_ranges(clip_a: vs.VideoNode,
...
     """
     if ranges is None:
         return clip_a
 
     if isinstance(ranges, str):  # type:ignore[unreachable]
         raise ValueError("replace_ranges: 'This function does not take strings! Please use a list of tuples or ints!")
 
+    nranges = normalize_ranges(clip_b, ranges)
+
+    if use_plugin and hasattr(core, 'remap'):
```
Comment (Setsugennoao) [inline lvsfunc/util.py:199]: There is a discrepancy in the function between use_plugin=True and False.
Namely, the mismatch; better way to handle it, if you want no mismatch is to set it to false and move the check at L205 before this call, if you want mismatch, replace the trim an splice calls that are done with python notation with std calls and mismatch=True

File: lvsfunc/util.py [hunk 11]
```diff
@@ -307,11 +260,6 @@ def force_mod(x: float, mod: int = 4) -> int:
     return mod ** 2 if x < mod ** 2 else int(x / mod + 0.5) * mod
 
 
-def clamp_values(x: float, max_val: float, min_val: float) -> float:
-    """Forcibly clamp the given value x to a max and/or min value."""
-    return min_val if x < min_val else max_val if x > max_val else x
-
-
 def get_neutral_value(clip: vs.VideoNode, chroma: bool = False) -> float:
```
Comment (Setsugennoao) [inline lvsfunc/util.py:263]: vsutil has this already

File: lvsfunc/util.py [hunk 12]
```diff
@@ -386,19 +336,32 @@ def get_coefs(curve: vs.TransferCharacteristics) -> Coefs:
...
+    """
+    Check for variable width or height and return an error if found.
+
+    :raises VariableResolutionError:    The clip has a variable resolution.
+    """
     if 0 in (clip.width, clip.height):
         raise VariableResolutionError(function)
 
 
 def check_variable(clip: vs.VideoNode, function: str) -> None:
```
Comment (Setsugennoao) [inline lvsfunc/util.py:358]: For all these check_* funcs, i suggest adding a typeguard, else they're pretty much useless for typing purposes and you'll still need to assert or whatever in your main function.
https://peps.python.org/pep-0647

File: requirements-dev.txt [hunk 13]
```diff
@@ -35,6 +35,7 @@ typed-ast==1.4.3
 typing-extensions==3.10.0.0
 types-setuptools>=57.4.17
 urllib3==1.26.9
-VapourSynth>=57
+VapourSynth>=58
 vsutil>=0.6.0
```
Comment (Setsugennoao) [inline requirements-dev.txt:39]: Stuff you've already got in requirements.txt has no reason to be in -dev requirements

File: requirements.txt [hunk 14]
```diff
@@ -1,5 +1,6 @@
-VapourSynth>=57
+VapourSynth>=58
```
Comment (Setsugennoao) [inline requirements.txt:1]: Bump up to R59, since other deps need that anyways

File: requirements.txt [hunk 15]
```diff
@@ -1,5 +1,6 @@
...
 rich>=8.0.0
 toolz>=0.9.0.1
 vs-kernels>=1.0.2
+vsrgtools>=0.1.3
 vsutil>=0.6.0
```
Comment (Setsugennoao) [inline requirements.txt:6]: Bump up to 0.8.0

File: lvsfunc/util.py [hunk 16]
```diff
@@ -622,18 +599,21 @@ def colored_clips(amount: int,
 
     Written by `Dave <mailto:orangechannel@pm.me>`_.
```
Comment (Setsugennoao) [inline lvsfunc/util.py:600]: Might want to put his github like for everyone else?
Mail is present there too.
https://github.com/OrangeChannel

**Review by Setsugennoao — COMMENTED** [review]
<3

File: lvsfunc/deblock.py (hunk 2 above)
Comment (LightArrowsEXE) [reply lvsfunc/deblock.py:158]: Clarity.

File: lvsfunc/deblock.py (hunk 4 above)
Comment (LightArrowsEXE) [reply lvsfunc/deblock.py:223]: Told you I'd forget

File: lvsfunc/dehalo.py (hunk 6 above)
Comment (LightArrowsEXE) [reply lvsfunc/dehalo.py:243]: They both have the same warnings in the docstring and when the user calls the function

File: lvsfunc/util.py (hunk 9 above)
Comment (LightArrowsEXE) [reply lvsfunc/util.py:59]: Probably next version. I want to double-check where it's being used first and make sure people change it lol

File: requirements-dev.txt (hunk 13 above)
Comment (LightArrowsEXE) [reply requirements-dev.txt:39]: mypy/flake8 only install this requirements.txt, not the other one.

File: lvsfunc/deblock.py (hunk 2 above)
Comment (Ichunjo) [reply lvsfunc/deblock.py:155]: I’d rather have `Catrom`.

File: lvsfunc/fun.py [hunk 17]
```diff
@@ -27,7 +27,7 @@ def minecraftify(clip: vs.VideoNode, div: float = 64.0, mod: int | None = None)
     """
     Transform your clip into a Minecraft.
 
-    Idea from Meme-Maji's Kobayashi memery (love you vardë).
+    Idea from Meme-Maji's Kobayashi memery (love you Vardë).
```
Comment (Ichunjo) [inline lvsfunc/fun.py:30]: Umu

File: lvsfunc/mask.py [hunk 18]
```diff
@@ -480,23 +483,24 @@ def mt_xxpand_multi(clip: vs.VideoNode,
...
-        clips += [m__imum(clips[-1], coordinates=[0, 0, 0, 1, 1, 0, 0, 0], **params)]
+        clips += [m__imum(clips[-1], coordinates=[0, 0, 0, 1, 1, 0, 0, 0], **params)]  # type:ignore[list-item]
 
     for x in range(end, int(end + sh - sw)):
-        clips += [m__imum(clips[-1], coordinates=[0, 1, 0, 0, 0, 0, 1, 0], **params).std.Limiter()]
+        clips += [m__imum(clips[-1], coordinates=[0, 1, 0, 0, 0, 0, 1, 0], **params)  # type:ignore[list-item]
+                  .std.Limiter()]
 
-    return clips
+    return clips  # type:ignore[return-value]
```
Comment (Ichunjo) [inline lvsfunc/mask.py:503]:
```suggestion
    clips: List[vs.VideoNode] = [clip]
    end = int(min(sw, sh)) + start

    for x in range(start, end):
        clips += [m__imum(clips[-1], coordinates=coordinates[x % 3], **params)]

    for x in range(end, int(end + sw - sh)):
        clips += [m__imum(clips[-1], coordinates=[0, 0, 0, 1, 1, 0, 0, 0], **params)]

    for x in range(end, int(end + sh - sw)):
        clips += [m__imum(clips[-1], coordinates=[0, 1, 0, 0, 0, 0, 1, 0], **params)
                  .std.Limiter()]

    return clips
```

File: lvsfunc/misc.py [hunk 19]
```diff
@@ -31,7 +31,7 @@
 
 def source(path: os.PathLike[str] | str, ref: vs.VideoNode | None = None,
            film_thr: float = 99.0, force_lsmas: bool = False,
-           tail_lines: int = 4, kernel: Kernel | str = Catrom(),
+           tail_lines: int = 4, kernel: Kernel | str = Bicubic(b=0, c=1/2),
```
Comment (Ichunjo) [inline lvsfunc/misc.py:34]: Pick one, 0.5 or 1/2

File: lvsfunc/misc.py [hunk 20]
```diff
@@ -71,14 +71,17 @@ def source(path: os.PathLike[str] | str, ref: vs.VideoNode | None = None,
                             If set above 100.0, it's silently lowered to 100.0 (Default: 99.0).
     :param force_lsmas:     Force files to be imported with L-SMASH (Default: False).
     :param kernel:          py:class:`vskernels.Kernel` object used for converting the `clip` to match `ref`.
-                            This can also be the string name of the kernel (Default: py:class:`vskernels.Catrom`).
+                            This can also be the string name of the kernel
+                            (Default: py:class:`vskernels.Bicubic(b=0, c=0.5)`).
```
Comment (Ichunjo) [inline lvsfunc/misc.py:75]: This too

File: lvsfunc/types.py [hunk 21]
```diff
@@ -8,7 +8,8 @@
 
 __all__: List[str] = [
     'Coefs', 'Coordinate', 'CreditMask', 'CURVES', 'CustomScaler', 'Direction', 'F', 'Matrix', 'Position', 'Range',
-    'RegressClips', 'Resolution', 'ScaleAttempt', 'SceneChangeMode', 'Size', 'T', 'VideoProp', 'VSFunction'
+    'RegressClips', 'Resolution', 'ScaleAttempt', 'SceneChangeMode', 'Size', 'T', 'VideoProp', 'VSFunction',
+    '_VideoNode'
```
Comment (Ichunjo) [inline lvsfunc/types.py:12]: You probably shouldn’t expose _VideoNode in the public API

File: lvsfunc/comparison.py [hunk 22]
```diff
@@ -598,14 +617,14 @@ def diff(*clips: vs.VideoNode,
     if thr >= 128:
         raise ValueError("diff: `thr` must be below 128!")
 
-    if clips and any(c.format is None for c in clips):
+    if clips and any(c.format is None for c in clips):  # noqa
```
[the remark this answers is not in the document]
Comment (LightArrowsEXE) [reply lvsfunc/comparison.py:620]:
```py
lvsfunc\comparison.py:620:8: SIM114 Use logical or ((clips and any(c.format is None for c in clips)) or (namedclips and any(nc.format is None for nc in namedclips.values()))) and a single body
    if clips and any(c.format is None for c in clips):
       ^
1     SIM114 Use logical or ((clips and any(c.format is None for c in clips)) or (namedclips and any(nc.format is None for nc in namedclips.values()))) and a single body
1
```

[the remark this answers is not in the document]
Comment (Setsugennoao) [reply lvsfunc/comparison.py:620]:
```suggestion
    if clips and not all(clips):
```

File: lvsfunc/deblock.py [hunk 23]
```diff
@@ -20,13 +20,13 @@
...
 def autodb_dpir(clip: vs.VideoNode, edgevalue: int = 24,
-                strs: Sequence[float] = [30, 50, 75],
+                strs: Sequence[float] = [10, 50, 75],
                 thrs: Sequence[Tuple[float, float, float]] = [(1.5, 2.0, 2.0), (3.0, 4.5, 4.5), (5.5, 7.0, 7.0)],
                 matrix: Matrix | int | None = None,
                 kernel: Kernel | str = Bicubic(b=0, c=0.5),
                 cuda: bool = True, write_props: bool = False,
                 **vsdpir_args: Any) -> vs.VideoNode:
-    """
+    r"""
```
[the remark this answers is not in the document]
Comment (LightArrowsEXE) [reply lvsfunc/deblock.py:29]: Prevents sphinx from doing weird fuckery to `text`\s.

File: lvsfunc/deblock.py [hunk 24]
```diff
@@ -165,52 +166,71 @@ def dpir(
     Converts to RGB -> runs DPIR -> converts back to original format, and with no subsampling if ``i444=True``.
     For more information, see `the original DPIR repository <https://github.com/cszn/DPIR>`_.
 
+    Thanks `Setsugen no ao <https://github.com/Setsugennoao>`_!
+
+    | Alias for this function is ``lvsfunc.vsdpir``.
```
[the remark this answers is not in the document]
Comment (LightArrowsEXE) [reply lvsfunc/deblock.py:171]: I don't know how to call a warning when users call a deprecated alias. I guess just create a function, call the warning in there, and then call the regular function?

**Comment by Ichunjo** [comment]
> @Ichunjo @Setsugennoao are there any other functions in (public) packages that I can replace in the `util` submodule? I think we're at the point where I'd rather depend on other packages that keep this updated than have to maintain my own utils.

`get_prop`, `get_neutral_value` and `get_matrix` could be in vsutil in my opinion. Now it’s a matter of providing a thesis to add them there :)

File: lvsfunc/deinterlace.py [hunk 25]
```diff
@@ -317,7 +333,8 @@ def deblend(n: int, clip: vs.VideoNode, rep: int | None) -> vs.VideoNode:
 def decomb(clip: vs.VideoNode,
            tff: bool | int = True, mode: int = 1,
            decimate: bool = True, vinv: bool = False,
-           rep: int | None = None, show_mask: bool = False,
+           rep: int | Sequence[int] = 0,
+           show_mask: bool = False,
            tfm_args: Dict[str, Any] = {},
            tdec_args: Dict[str, Any] = {},
            vinv_args: Dict[str, Any] = {},
```
[the remark this answers is not in the document]
Comment (LightArrowsEXE) [reply lvsfunc/deinterlace.py:340]: The issue with this is mainly that you quickly run into `TypeError: 'NoneType' object is not a mapping`, requiring a lot of additional lines just to handle those.

File: lvsfunc/comparison.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (Setsugennoao) [reply lvsfunc/comparison.py:620]: Tfw you forgot to write half of the snippet.
It doesn't throw any errors for me

File: docs/requirements.txt [hunk 26]
```diff
@@ -26,4 +26,5 @@ sphinxcontrib-htmlhelp==2.0.0
 sphinxcontrib-jsmath==1.0.1
 sphinxcontrib-qthelp==1.0.3
 sphinxcontrib-serializinghtml==1.1.5
+typing-extensions==3.10.0.0
```
[the remark this answers is not in the document]
Comment (Ichunjo) [reply docs/requirements.txt:29]: Oh it’s the reqs for docs. Well I guess it should be fine but there are too many dependencies. I’m pretty sure most of them are in the dependence’s tree of one of them

[the remark this answers is not in the document]
Comment (LightArrowsEXE) [reply docs/requirements.txt:29]: Possibly, but I'm also wondering if it's actually worth untangling them tbh
